### ISSUE TITLE
LTR performance in version 1.7 not reproducible in 3.0

### ISSUE THREAD

**Issue by emdodds** [dmlc/xgboost#11635]
I'd like to upgrade xgboost to the latest version to take advantage of new features, but in my learning-to-rank problem I observe a degradation in performance when I upgrade. I appreciate the suggestions in the docs [here](https://xgboost.readthedocs.io/en/stable/tutorials/learning_to_rank.html#comparing-results-with-version-1-7) on how to obtain "similar result" and have implemented them but the difference in performance persists. The difference is meaningful and seems too large to be due to different random seeds or floating point arithmetic.

I can't share the real problem I'm working on, but it's easy to observe this on synthetic data. Here is an example:

```python
import numpy as np
import xgboost as xgb
from scipy.stats import spearmanr
from sklearn.model_selection import train_test_split


def generate_synthetic_data(
    n_examples: int = 10000,
    n_features: int = 50,
    noise_level: float = 13,
    seed: int = 42) -> tuple[np.ndarray, np.ndarray]:
    rng = np.random.RandomState(seed)

    X = np.zeros((n_examples, n_features), dtype=np.float32)

    # Different types of features
    X[:, :20] = rng.normal(size=(n_examples, 20))
    X[:, 20:45] = rng.lognormal(0, 1, (n_examples, 25))
    X[:, 45:50] = X[:, :5] * X[:, 5:10]

    # Synthetic labels
    linear_weights = np.zeros(n_features)
    linear_weights[:10] = rng.normal(size=10) * 2
    linear_weights[10:20] = rng.normal(size=10) * 0.5

    y = X @ linear_weights
    y += 0.5 * (X[:, 0] ** 2 - 1)
    y += 0.3 * (X[:, 1] * X[:, 2])
    y += np.where(X[:, 20] > 2, 1.5, 0)
    y += np.where(X[:, 30] == 1, -0.8, 0)
    y += rng.normal(size=n_examples) * noise_level

    return X, y


def train_and_eval():
    X, y = generate_synthetic_data()
    Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.2, random_state=42)

    if int(xgb.__version__.split('.')[0]) > 1:
        params = {
            "objective": "rank:pairwise",
            "lambdarank_pair_method": "mean",
            "lambdarank_score_normalization": False,
            "base_score": 0.5,
            "tree_method": "approx",
            "lambdarank_num_pair_per_sample": 1,
        }
    else:
        params = {
            "objective": "rank:pairwise",
        }

    model = xgb.XGBRanker(**params, random_state=42)

    qid = np.zeros(len(Xtrain))

    model.fit(Xtrain, ytrain, qid=qid)

    ypred = model.predict(Xtest)

    spearman_corr, _ = spearmanr(ytest, ypred)

    print("xgb version:", xgb.__version__)
    print("Spearman coefficient:", spearman_corr)

if __name__ == "__main__":
    train_and_eval()
```

Results running from two different venvs with different xgboost versions:

```
xgb version: 3.0.4
Spearman coefficient: 0.3377999944499986
```

```
xgb version: 1.7.6
Spearman coefficient: 0.34664375066093767
```

Is this magnitude of difference expected? Tweaking the data, I can get differences in either direction, but of course I don't have that luxury with my real world problem where I consistently observe worse performance with version 3.0.4 even after hyperparameter tuning. 

If there is something I'm missing, please let me know. I recognize full data ranking is not the usual use case for LTR but it's what I'm using it for.

**Comment by trivialfis** [comment]
Thank you for raising the issue. I will look into it.

**Comment by trivialfis** [comment]
Could you please consider using the `hist` tree method instead? In many cases, it can obtain faster computation performance and higher accuracy. There are two differences between these two versions after setting the documented parameter set:
- The approx tree method was slightly modified to match the behaviour of hist for the quantization process.
- The seed was changed.

I did some tests locally by redefining the seed and using the `hist` tree method for both versions, the output models are the same.

In addition, I opened a PR to improve the seed https://github.com/dmlc/xgboost/pull/11639 .

**Comment by emdodds** [comment]
Thanks so much for the quick response!

Yes, I use the `hist` tree method generally, I just used `approx` for my toy example so I could follow the instructions in the doc exactly. Setting `tree_method` to `hist` for both versions does not resolve my problem. 

I am not able to reproduce your finding that the models are the same when using version 1.7.6 and the version in your PR. I modified my little test script slightly to use `hist` and also print the mean of the data to verify only the models are different:

```python
import numpy as np
import xgboost as xgb
from scipy.stats import spearmanr
from sklearn.model_selection import train_test_split


def generate_synthetic_data(
    n_examples: int = 10000,
    n_features: int = 50,
    noise_level: float = 13,
    seed: int = 42) -> tuple[np.ndarray, np.ndarray]:
    rng = np.random.RandomState(seed)

    X = np.zeros((n_examples, n_features), dtype=np.float32)

    # Different types of features
    X[:, :20] = rng.normal(size=(n_examples, 20))
    X[:, 20:45] = rng.lognormal(0, 1, (n_examples, 25))
    X[:, 45:50] = X[:, :5] * X[:, 5:10]

    # Synthetic labels
    linear_weights = np.zeros(n_features)
    linear_weights[:10] = rng.normal(size=10) * 2
    linear_weights[10:20] = rng.normal(size=10) * 0.5

    y = X @ linear_weights
    y += 0.5 * (X[:, 0] ** 2 - 1)
    y += 0.3 * (X[:, 1] * X[:, 2])
    y += np.where(X[:, 20] > 2, 1.5, 0)
    y += np.where(X[:, 30] == 1, -0.8, 0)
    y += rng.normal(size=n_examples) * noise_level

    return X, y


def train_and_eval():
    X, y = generate_synthetic_data()
    Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.2, random_state=42)

    if int(xgb.__version__.split('.')[0]) > 1:
        params = {
            "objective": "rank:pairwise",
            "lambdarank_pair_method": "mean",
            "lambdarank_score_normalization": False,
            "base_score": 0.5,
            "tree_method": "hist",
            "lambdarank_num_pair_per_sample": 1,
        }
    else:
        params = {
            "objective": "rank:pairwise",
            "tree_method": "hist",
        }

    model = xgb.XGBRanker(**params, random_state=42)

    qid = np.zeros(len(Xtrain))

    model.fit(Xtrain, ytrain, qid=qid)

    ypred = model.predict(Xtest)

    spearman_corr, _ = spearmanr(ytest, ypred)

    print("xgb version:", xgb.__version__)
    print("Data means:", np.mean(X), np.mean(y))
    print("Spearman coefficient:", spearman_corr)

if __name__ == "__main__":
    train_and_eval()
```

Running this script from my old venv with 1.7.6 installed I get

```
xgb version: 1.7.6
Data means: 0.8237397 0.20761291450653702
Spearman coefficient: 0.325826371956593
```

whereas running from another venv where I installed from `75e53e384` (the state of your PR) I get

```
xgb version: 3.1.0-dev
Data means: 0.8237397 0.20761291450653702
Spearman coefficient: 0.32591785247946314
```

It's possible I messed up the installation, but this is different from what I get in 3.0.4 so I do think I picked up your change to the first iteration seed:

```
xgb version: 3.0.4
Data means: 0.8237397 0.20761291450653702
Spearman coefficient: 0.3125582581395645
```

The difference between 1.7.6 and the PR version looks smaller, but tweaking the data parameters I can easily get larger differences again so I think the closer match is just a coincidence.

Am I missing something in how you got matching results? When you say "the output models are the same" do you literally mean that the underlying trees and weights are identical or do you mean the performance is very close?

**Comment by trivialfis** [comment]
Thank you for testing it. It won't be the same, with or without my PR.

The old implementation in 1.7 has the seed `(iter + 1) *1111` for all query groups.  Once I changed the seed back to this old implementation along with the tree method, the result matches between versions (Used in the same place as the change my PR made `std::uint32_t seed = ...`).

I changed the seed after 1.7 to account for query group index to make the sampling a bit more random (not necessarily true, since different groups generate different pairs regardless of the seed, but helpful in practice as evident by your example). The difference you observed after switching to hist, is indeed coincidence caused by the seed.

My PR simply makes the first iteration have different seeds for different groups, which does improve the results as previously suggested. I don't intend to change back to the same seed used by 1.7 unless there's a good reason to.

**Comment by trivialfis** [comment]
That said, you are more than welcome to check whether changing the seed back can help matching the results. I can create a patch tomorrow if you haven't already.

**Comment by emdodds** [comment]
Thanks for clarifying. I tried changing the seed back to `(iter + 1) *1111` and I am able to _sometimes_ match results between 1.7.6 and my patch off of 3.0.4 with that change. They match on some datasets and not others.

I haven't been able to figure out when they match and when they don't. My example script I pasted above gives the same results as is, but if I double the size of the synthetic data the results differ substantially again. So I think there is something else besides this seed that differs between the versions.

While the ability to exactly replicate 1.7.6 models in 3.x would be nice, it would also be good enough for me if the performance of 3.x converged or was better for large datasets, or in some reasonable hyperparameter regime.

**Comment by trivialfis** [comment]
Let me dig deeper.

**Comment by trivialfis** [comment]
There are some floating-point differences in the hist tree method due to different parallel execution strategies. This is random noise generated by the limitations of floating point. We can close this issue now.
